From bab4af92ad0c732070bf255937f57bfd9084c909 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Mon, 28 Sep 2020 21:27:12 +0300
Subject: [PATCH] [SW-21029] ppboot: goya2_arc: copy data vars from FWSRAM

In secure-boot mode of Goya2, ppboot runs from FW SRAM
that is read-only. In order to facilitate usage of
initialized global variables, they have to be relocated
from FWSRAM to DCCM so that they can be updated. Hence
this patch copies such variables before jumping to
ppboot_main

Change-Id: Id0482a437cb10ad6d624f2797d9c0dcf0a16ede5
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/ppboot/crt0_ppboot.S b/ppboot/crt0_ppboot.S
index e4b2e30..66f144b 100644
--- a/ppboot/crt0_ppboot.S
+++ b/ppboot/crt0_ppboot.S
@@ -24,6 +24,19 @@
 	lpnz    1f
 	st.ab   0, [r5, 4]
 1:
+	/* copy data and sdata sections */
+	mov_s	r1, @_erodata		; r1 = end of the text/ro section
+	mov_s	r2, @_data
+	sub	r3, @_esdata, r2	; r3 = size of total data in bytes
+
+	asr_s	r13, r3
+	asr.f	lp_count, r13
+
+	lpnz	2f
+	ld.ab	r5,[r1,4]
+	st.ab	r5,[r2,4]
+2:
+
 #endif
 
 #ifdef __aarch64__
